academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+
+        upsert_files = {}
+        for datafile in upsert_files_to_merge:
+            transform_file = datafile.transform_file_path(entity_type=self.entity_list["upsert"].type)
+
+            with open(transform_file, "r") as f_in:
+                incoming_data = [json.loads(line) for line in f_in]
+
+            # Get list of 'primary keys' for each record.
+            records = [str(record[self.entity_list["upsert"].pmid_location]["PMID"]) for record in incoming_data]
+
+            set_new_records_to_check = set(records)
+            for key_file_path, records_in_file in upsert_files.items():
+                # Find the primary keys that are the same in the old updatefile.
+                set_old = set(records_in_file)
+                same_records = set_old.intersection(set_new_records_to_check)
+
+                # Remove primary keys in the old upsert file set if it is present in a newer upsert file.
+                set_old.difference_update(same_records)
+
+                if same_records:
+                    logging.info(
+                        (
+                            f"Found {len(same_records)} newer records in file: {transform_file.split('/')[-1]} vs {key_file_path.split('/')[-1]}."
+                            f" {len(set_old)} records left in {key_file_path.split('/')[-1]}"
+                        )
+                    )
+
+                # Relace with list of primary keys with old keys removed.
+                upsert_files[key_file_path] = list(set_old)
+
+            # Add newest updatefile to list of updatefiles and it's records.
+            upsert_files[transform_file] = records
+
+        merged_upsert_files = []
+        first_file_index = upsert_files_to_merge[0].file_index
+        last_file_index = upsert_files_to_merge[-1].file_index
+        for file_count, (transform_file, records_to_write) in enumerate(upsert_files.items()):
+            logging.info(f"Reading in file to pull out records - {transform_file}")
+            with open(transform_file) as f_in:
+                incoming_data = [json.loads(line) for line in f_in]
+
+            # Turn list of records to a dictionary for faster lookup.
+            records = {
+                str(record[self.entity_list["upsert"].pmid_location]["PMID"]): record for record in incoming_data
+            }
+
+            merged_output_file = datafile.merged_transform_file_path(
+                self.entity_list["upsert"].type, first_file_index, last_file_index, part_num=file_count + 1
+            )
+
+            logging.info(f"Writing {len(records_to_write)} records to file - {merged_output_file}")
+
+            # Write to compressed *.jsonl.gz as this will be uploaded to GCS and imported into BQ.
+            with gzip.open(merged_output_file, "w") as f_out:
+                for record_key in records_to_write:
+                    f_out.write(str.encode(json.dumps(records[record_key], cls=PubMedCustomEncoder) + "\n"))
```

Comment:
You could make a function `save_pubmed_jsonl` and use this in multiple places:

```suggestion
            save_pubmed_jsonl(merged_output_file, [records[record_key] for record_key in records_to_write])
```

For instance:
```python
def save_pubmed_jsonl(file_path: str, data: List[dict]):
    if file_path.endswith(".gz"):
        with gzip.open(file_path, "w") as f_out:
            for line in data:
                f_out.write(str.encode(json.dumps(line, cls=PubMedCustomEncoder) + "\n"))
    else:
        with open(file_path, "w") as f_out:
            for line in data:
                f_out.write(json.dumps(line, cls=PubMedCustomEncoder) + "\n")
```